common/logging/src/tracing_logging_layer.rs

```diff
@@ -0,0 +1,126 @@
+use std::collections::HashMap;
+use std::fs::{self, create_dir_all, File, OpenOptions};
+use std::io::Write;
+use std::path::PathBuf;
+use std::sync::mpsc::{self, Sender};
+use std::thread;
+
+lazy_static! {
+    pub static ref TRACING_LOGGING_DEPENDENCIES: Vec<String> =
+        vec!["libp2p_gossipsub".to_string(), "discv5".to_string()];
+}
+
+/// Layer that handles `INFO`, `WARN` and `ERROR` logs emitted per dependency and
+/// writes them to a file. Dependencies are enabled via the `RUST_LOG` env flag.
+pub struct LoggingLayer {
+    pub file_writer_streams: HashMap<String, NonBlockingFileWriter>,
+}
+
+impl<S: tracing_core::Subscriber> tracing_subscriber::layer::Layer<S> for LoggingLayer {
+    fn on_event(
+        &self,
+        event: &tracing_core::Event<'_>,
+        _ctx: tracing_subscriber::layer::Context<'_, S>,
+    ) {
+        let meta = event.metadata();
+
+        let target = match meta.target().split_once("::") {
+            Some((crate_name, _)) => crate_name,
+            None => "unknown",
+        };
+
+        let Some(file_writer) = self.file_writer_streams.get(target) else {
+            return;
+        };
+
+        let mut visitor = LogMessageExtractor {
+            message: String::default(),
+        };
+
+        event.record(&mut visitor);
+
+        let _ = file_writer.write(visitor.message);
+    }
+}
+
+pub struct NonBlockingFileWriter {
+    sender: Sender<String>,
+}
+
+impl NonBlockingFileWriter {
+    pub fn new(path: &std::path::Path, max_file_size: u64) -> Result<Self, std::io::Error> {
+        let (sender, receiver) = mpsc::channel();
+        let path = path.to_path_buf();
+
+        thread::spawn(move || {
+            if !path.exists() {
+                let mut dir = path.clone();
+                dir.pop();
+
+                // Create the necessary directories for the correct service and network.
+                if !dir.exists() {
+                    let res = create_dir_all(dir);
+
+                    match res {
+                        Ok(_) => (),
+                        Err(e) => {
+                            eprintln!("Failed to create dir: {:?}", e);
+                            return;
+                        }
+                    }
+                }
+            }
+
+            let mut file = match OpenOptions::new().create(true).append(true).open(&path) {
+                Ok(file) => file,
+                Err(e) => {
+                    eprintln!("Failed to open file: {:?}", e);
+                    return;
+                }
+            };
+
+            for message in receiver {
```

Comment:
> Shouldn't we have a loop here somewhere. It seems like this thread just reads all the messages, writes them, then ends?

As far as i understand it, the mpsc channel creates a sender and receiver. The receiver blocks until a message is available (as long as sender still exists).

https://doc.rust-lang.org/std/sync/mpsc/fn.channel.html

So that `for message in receiver` loop blocks until a message is received and should never end as long as the sender exists.

> It also looks like there is one of these threads per dependency/file. I was expecting an await, or perhaps this tasks continually gets created, but I only saw one initialisation. I'm likely missing something.

There is one thread per dependency, the stream initialization per dependency happens [here](https://github.com/sigp/lighthouse/pull/4988/files#diff-51538e097658c4a146e32787ee886030ee7abb4cacb1e8a873d8d5030923117aR240-R247)

Might be a bit overkill to spawn a new stream per dependency, I can change it to one stream for all dependencies if that makes more sense

Reply:
ah yeah all good. Sorry have been in tokio async land too long and forgot about std channels :p